Add support for jsx-runtime

JacopoPatroclo opened the pull request:
It would be nice to support the auto-import of the jsx runtime, removing the need to import the `@kitajs/html/register` sub-package.

This PR aims to do that.

@arthurfiorette let me know if this seems like an improvement and if you would like to make this the standard way of configuring the tsconfig for this library. In that case, I will adapt the docs.

arthurfiorette commented:
Hey, can you give more details about what and why you're doing this?

JacopoPatroclo commented:
For sure.

## Current status
If you want to use this library in a typescript project you must use a tsconfig that looks something like this
```json
{
  "compilerOptions": {
    "jsx": "react",
    "jsxFactory": "Html.createElement",
    "jsxFragmentFactory": "Html.Fragment",
    "plugins": [{ "name": "@kitajs/ts-html-plugin" }]
  }
}
```
the property "jsx": "react" tells typescript to substitute the jsx elements with a call to "Html.createElement" (or "Html.Fragment" in case there is a fragment). This implies that you have to have in your code a declaration of the Html object. The choice is to import it per file or globally using the module `@kitajs/html/register` as (recommended by the docs).

## The catch
When you choose the global import as a way to import the Html object you are going to encounter issues when you are trying to unit test your code (that uses jsx somewhere). This is because you have to import the "register module" on top of every unit test file.
Another case could be made about the pollution of the global object, but I'm not going into detail about that.

## Solution
Typescript support "jsx": "react-jsx" that will transpile the jsx code with a call to a function called `jsx` or `jsxs` that will be automatically imported from a module named what you configure in `jsxImportSource` concatenating `jsx-runtime`. Example:

```json
{
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "@kitajs/html",
    "plugins": [{ "name": "@kitajs/ts-html-plugin" }]
  }
}
```
code
```typescript
const a = <h1>Hello</h1>
```
transpilled to something like

```javascript
const runtime = require('@kitajs/html/jsx-runtime')
const a = runtime.jsx('h1', { children: 'Hello' })
```
As you can see there is no global pollution and the import to the jsx function is made by the transpiler; removing the need to have a global import to `@kitajs/html/register` module.

## The implementation
I've followed the specification expected by typescript during the implementation of the jsx-runtime sub module.
I've created a ts configuration to test that all the current tests work using this different way of configuring typescript/jsx. I had to add "noUnusedLocals": false because all the test import the Html object that is no longer needed with this new config.

Hope this is clearer now, let's discuss about it!

Docs:
https://www.typescriptlang.org/tsconfig#jsx
https://github.com/facebook/react/blob/59831c98cffe0edf706238b067928e7cf54d1065/packages/react/jsx-runtime.js#L4
https://github.com/preactjs/preact/blob/a003d429f4cfa6c131f01eab52d556b242b3fc59/jsx-runtime/src/index.js#L4

arthurfiorette commented on README.md:
```diff
@@ -191,10 +207,9 @@ issues in your code editor and enhance your code's security.
-> Please utilize our `@kitajs/ts-html-plugin` to emit TypeScript errors wherever you are
-> exposed to XSS. Refer to [Getting Started](#getting-started) for installation
-> instructions.
+> [!IMPORTANT] Please utilize our `@kitajs/ts-html-plugin` to emit TypeScript errors
```
your formatter broke all github flavored markdown warnings.

arthurfiorette commented on README.md:
```diff
@@ -373,8 +388,7 @@ const html = renderToStream(renderUserPage);
 
-> [!NOTE]  
-> The `renderToStream()` is returns a native node/bun stream, head over to our
+> [!NOTE] The `renderToStream()` is returns a native node/bun stream, head over to our
```
here too

arthurfiorette commented on README.md:
```diff
@@ -485,10 +499,9 @@ error is thrown.
-> Until [#14729](https://github.com/microsoft/TypeScript/issues/14729) gets implemented,
-> you need to manually cast `JSX.Element` into strings if you are sure there is no inner
-> async components in your component tree.
+> [!NOTE] Until [#14729](https://github.com/microsoft/TypeScript/issues/14729) gets
```
here too

arthurfiorette commented on README.md:
```diff
@@ -678,9 +691,8 @@ component. This does not support unclean components / props processing.
-> [!WARNING]  
-> This feature is a special use case for rendering **entire page templates** like what you
-> would do with handlebars or nunjucks.
+> [!WARNING] This feature is a special use case for rendering **entire page templates**
```
here too

arthurfiorette commented on jsx-runtime.d.ts:
```diff
@@ -0,0 +1,15 @@
+
+export function safeChildrenArray(children: Children): Array<Children>;
+
+export function jsx(
```
Re-exporting them would force double maintaince, probably this syntax is better:

```ts
export {
  Fragment,
  createElement as jsxs,
  createElement as jsx
} from './index';
```

JacopoPatroclo replied:
Unfortunately the jsx and jsxs function signature is different from the createElement one. We can't do that, hence the tests using the `tsconfig.jsxruntime.json` ts config. Those tests, in my opinion, are necessary to ensure we don't break the contract between our implementation and the expected one by typescript.

arthurfiorette commented on jsx-runtime.js:
```diff
@@ -0,0 +1,43 @@
+    // must be provided as sequential arguments to createElement
+    // so in order to spread it safely we need to ensure that it is an array
+    const childrenArray = safeChildrenArray(children);
+    return createElement(type, props, ...childrenArray);
```
`createElement` childrens can be a infinite depth array, or a single item. Because of the below implementation, `contentsToString` flatterns on the go the provided array and simply work.

https://github.com/kitajs/html/blob/d556fcddfff0337b1f0ec0b8b2a81f1069d45155/index.js#L395-L399

JacopoPatroclo replied:
I'm not sure that I follow you on this one. The issue is that the createElement function expects children to be passed as sequential parameters after the type of tag and the props. Maybe we should change the signature of the createElement function to support both cases. But again, I don't think I have understood your point here.

arthurfiorette commented:
When importing other things from `@kitajs/html`, should we still require the user to create the import just like when importing `useState` from React, or should we keep the `/register`-way that, once imported, `Html` namespace is available everywhere?

Reading more about `jsx-runtime`, I think we should completely remove the `/register` import and push the usage of `jsx-runtime` when global usage is wanted.

arthurfiorette commented on package.json:
```diff
@@ -10,13 +10,16 @@
+    "exec-test": "c8 --reporter lcov --reporter text node --trace-warnings --enable-source-maps --test dist/test",
     "format": "prettier --write .",
-    "test": "tsc && c8 --reporter lcov --reporter text node --trace-warnings --enable-source-maps --test dist/test",
+    "test": "(tsc && pnpm exec-test) && (tsc -p ./tsconfig.jsxruntime.json && pnpm exec-test)",
```
This tests code, but not ours. When double transpiling with react-jsx, the only thing being tested is if `tsc` is traspiling correctly, which I'm sure microsoft/typescript has better tests than these.

Instead, you could create a test suite to ensure the file `/jsx-runtime.js` is well typed and exports `jsx`, `jsxs` and `Fragment`.

arthurfiorette replied:
please also remove tsconfig.jsxruntime.json and exec-test script

JacopoPatroclo replied:
See https://github.com/kitajs/html/pull/129#discussion_r1495005802

arthurfiorette commented on jsx-runtime.js:
```diff
@@ -0,0 +1,43 @@
+ *   children: import('./index.d.ts').Children
+ * ) => import('./index.d.ts').Children[]}
+ */
+function safeChildrenArray(children) {
```
```suggestion
```

arthurfiorette commented on README.md:
```diff
@@ -167,7 +182,8 @@ to generate HTML. Here are two options for importing the `@kitajs/html` package:
 2. **Use the register to add a global namespace**: Import the `register` to globally
-   register all necessary functions for convenience.
+   register all necessary functions for convenience. You don't need to do that if you are
+   using the `jsxImportSource` option in your `tsconfig.json`.
```
Should we prefer /register import or `react-jsx` way of "globally" installing kitajs/html? I'm not sure if both would be good.

JacopoPatroclo replied:
I think that the best approach is to follow how preact do it https://preactjs.com/guide/v10/typescript/. They let the user choose how to configure Typescript between importing the h and Fragment functions into every file or using the "react-jsx" + "jsxImportSource" approach (automatic import done by the transpiler).

Side note here:
We should also consider adding some documentation on how to setup a project using babel or similar for non-typescript users.

arthurfiorette commented:
I've seen [`jsxDEV`](https://github.com/preactjs/preact/blob/a003d429f4cfa6c131f01eab52d556b242b3fc59/jsx-runtime/src/index.js#L181) and `__jsxDEV` exports in old react project sources, should we also provide them?

I knew about `react-jsx` before but could not find any actual spec or implementation description to follow a standard here. That's why I never followed the path and implemented it... Is there something we could validate against?

JacopoPatroclo commented:
> I've seen [`jsxDEV`](https://github.com/preactjs/preact/blob/a003d429f4cfa6c131f01eab52d556b242b3fc59/jsx-runtime/src/index.js#L181) and `__jsxDEV` exports in old react project sources, should we also provide them?
> 
> I knew about `react-jsx` before but could not find any actual spec or implementation description to follow a standard here. That's why I never followed the path and implemented it... Is there something we could validate against?

I don't think that this package need __jsxDEV because it does not have an actual need for it. The __jsxDEV is used to run checks during development on, for example, the fact that you are mapping elements without providing a key prop. Or other stuff like that.

I had the same problem trying to create this first draft. The implementation that I did was against the tests using the alternative typescript config. There is a lot of React-related content on the topic of jsx on the internet but very little on "server side" jsx runtimes. I guess that we should start writing our stuff driven by trial and error.

JacopoPatroclo commented:
> Additionally, we could drastically improve kitajs/html performance with support for precompiling JSX.
> 
> Read more:
> 
> [oven-sh/bun#6878](https://github.com/oven-sh/bun/issues/6878) https://deno.com/blog/v1.38#fastest-jsx-transform https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html https://github.com/reactjs/rfcs/blob/createlement-rfc/text/0000-create-element-changes.md#motivation

I'm on it

arthurfiorette commented:
> Additionally, we could drastically improve kitajs/html performance with support for precompiling JSX.
> 
> Read more:
> 
> [oven-sh/bun#6878](https://github.com/oven-sh/bun/issues/6878) https://deno.com/blog/v1.38#fastest-jsx-transform https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html https://github.com/reactjs/rfcs/blob/createlement-rfc/text/0000-create-element-changes.md#motivation

It looks like this is still only supported in Deno + Preact, it's a good idea, but it won't be used by anyone until support arrives later in TS or in packagers. I created a proposal at microsoft/TypeScript#57468 for this, but I'm not sure if it will be accepted by the community...

Until then, this feature will have to wait.

arthurfiorette commented:
We need to keep the same performance for `createElement` and `jsx`/s. Thus, having more functions calls drastically decreased performance, thus I reimplemented `createElement` for `jsxs` (children: Children[]) and `jsx` (children?: Children).

arthurfiorette commented:
I also simplified JSDoc comments for createElement functions, as their typings were wrong.

arthurfiorette commented:
Current jsx implementation is still slower

![image](https://github.com/kitajs/html/assets/47537704/023af9a7-dda9-4af7-94f9-93da14e33fde)

arthurfiorette commented:
`jsx` is only used with no children or a single one, this allows us to replace the `contentsToString` call which has a loop to a single if/else branch.

![image](https://github.com/kitajs/html/assets/47537704/823d4f17-e08c-41e9-91b7-9b022b3a2e7d)

I also improved the original `contentsToString` because it previously didn't serialized `false` and `NaN`, which was probably a bug.

arthurfiorette commented:
With `react-jsx`, we can safely type the `Children` attribute, previously it was `any`. This removes even more checks inside `contentsToString` and move them to the TSC compiler:

```tsx
<div>{{a:1}}</div> // Previously was fine and was transformed to [object Object], which is unintended
```

now:

```tsx
// ts(2353): Object literal may only specify known properties, and 'a' does not exist in type 'Promise<Children> | Children[]'
<div>{{a:1}}</div>
```

Increasing even further the performance

arthurfiorette commented:
`jsx-runtime` is faster than `createElement` _(That's probably why react itself moved to it)_. I think we should make it default and deprecate the `createElement` way.

arthurfiorette commented:
Current tasks:

- [ ] jsx and jsxs jsdoc
- [ ] Move documentation to be `react-jsx` first
- [ ] Fix serialization test (false now is `"false"` instead of `""`...)
- [ ] Add a table containing input -> outputs to be expected in the readme, like `true -> "true"`, `[1,"3"] -> "13"` and so on... _(Functions and objects will throw tsc errors)_
- [ ] Ensure everything keeps working with `ts-html-plugin`.
- [ ] Add a way to deprecate older `createElement`/`import Html` way

arthurfiorette commented:
I improved even further JSX/JSXS performance, using the average of kb/html and the time per iteration of our benchmarks, this is the result:

![image](https://github.com/kitajs/html/assets/47537704/03f705f1-4bd9-4464-9225-931c44ada488)

***Holy shit 1.2gbps*** 

I can finally rest in peace now XD.

---

https://github.com/fastify/process-warning is a great example on how to deprecate things, so I followed their pattern (without installing it as a dependency). 

This is how it shows up on bun and nodejs:

![image](https://github.com/kitajs/html/assets/47537704/da95665c-572f-45dd-bde5-c1850756fd70)

And this is what the user will see after clicking the link:

![image](https://github.com/kitajs/html/assets/47537704/ebe2fff4-01cd-4dfb-b46a-a2ee947b4187)

---

I also added more examples to the serialization table.

arthurfiorette commented:
Could you finish these two tasks below?

- [ ] Ensure everything keeps working with `ts-html-plugin`.
- [ ] Fix benchmarks to use newer jsxs syntax for kita

arthurfiorette commented:
@JacopoPatroclo could you finish these two tasks below?

- [ ] Ensure everything keeps working with `ts-html-plugin`.
- [ ] Fix benchmarks to use newer jsxs syntax for kita

> @arthurfiorette let me know your thoughts about the deprecation message I add to the register module. Moreover, I think that adding the @deprecated jsdoc property to the createElement function does not make a lot of sense since it is usually used by the transpiler and not directly by the end user. We should make it explicit in the docs and at some point in the future substitute the functionality with a throw Error that explains how to migrate and where to find the documentation to do it.

I think only using the correct way to deprecate it (process warning) is enough for now, let's see how the users react to it.

> I will also add to the tasks:
> * [ ] Migrate tests to utilize the jsx-runtime instead of the createElement function (this will increase the code coverage also I think)

Awesome!

kurtextrem commented on README.md:
```diff
@@ -985,6 +1036,29 @@ Which, when called, returns this string:
+| `<div>{false}</div>`       | `<div>false</div>`    |
+| `<div>{null}</div>`        | `<div></div>`         |
+| `<div>{undefined}</div>`   | `<div></div>`         |
+| `<div>{[1, 2, 3]}</div>`   | `<div>12</div>`       |
```
the output here seems off, or at least odd?

arthurfiorette replied:
thanks!

JacopoPatroclo commented:
@arthurfiorette I'm encountering issues migrating the benchmarks to use the new JSX-runtime.
The main pinpoint is related to the fact that the benchmarks are written using es-modules but the package uses commonjs. 
This should not be an issue but the jsx runtime import is transpilled by typescript like this:

```js
import { jsx as _jsx, jsxs as _jsxs } from '../../jsx-runtime'
```

but as stated from the Node.js docs you can only import common-js modules using a "global import"

```
import something from 'common-js-module'
```

I've tried to convert everything to common-js but the package that we are using for benchmarking (`mitata`) seems to only work when imported using esm syntax. It's kind of a mess. I'm not sure how to bring home this refactor.

Side note, we should consider using pnpm-workspace to handle the benchmarks piece, it would be nice to reference the jsx-runtime using the real path (`@kitajs/htm`) instead of something like (`../..`). This could be achieved by creating a project that has between it's dependencies `@kitajs/htm` with the version `workspace:*`

Let me know what you think of my reasoning above and about the workspace approach.

arthurfiorette commented:
> Let me know what you think of my reasoning above and about the workspace approach.

You could try to self install using a file path at `benchmark/package.json`:

```json
{
  "type": "module",
  "dependencies": {
    "kitajs": "file:../.." 
  }
}
```

---

Regarding the workspace approach, I'll probably create a workspace to group all html related projects into a single repository:

https://github.com/kitajs/html
https://github.com/kitajs/fastify-html-plugin
https://github.com/kitajs/ts-html-plugin

But I do not have time for this now, maybe after this PR gets merged I can work on it.

JacopoPatroclo commented:
@arthurfiorette let me know if my refactoring of the project suits you.
I've separated each renderer into its package, now we have a cleaner way to handle the ts compilation for each package.
The benchmark runner has the same checks and bench as before to ensure that each renderer is going to output the same markup as `@kitajs/html`.

JacopoPatroclo commented:
> > Let me know what you think of my reasoning above and about the workspace approach.
> 
> You could try to self install using a file path at `benchmark/package.json`:
> 
> ```json
> {
>   "type": "module",
>   "dependencies": {
>     "kitajs": "file:.." 
>   }
> }
> ```
> 
> Regarding the workspace approach, I'll probably create a workspace to group all html related projects into a single repository:
> 
> https://github.com/kitajs/html https://github.com/kitajs/fastify-html-plugin https://github.com/kitajs/ts-html-plugin
> 
> But I do not have time for this now, maybe after this PR gets merged I can work on it.

I can help with this task, I'm very familiar with [Nx](https://nx.dev/) and pnpm workspaces.

arthurfiorette commented:
Can you also merge from main? pnpm should resolve conflicts for us

arthurfiorette commented:
Since we're already here, I think its safe to move `ts-html-plugin` and `fastify-html-plugin` to this repository. I'll let the monorepo setup similar to how its configured in https://github.com/kitajs/kitajs.

arthurfiorette commented:
Just moved both packages into this branch. Could you review it?

Tasklist:

- [ ] Update links and references inside all package jsons.
- [ ] Ensure tests are working on all packages
- [ ] Test CI/CI
- [ ] Test changeset
- [ ] Update docs and reference links
- [ ] Do a simple publish --dry-run to ensure all files are being bundled correctly
- [ ] See if there's no missing or left configuration files
- [ ] Update monorepo readme's

arthurfiorette commented:
Also, our current benchmark isn't fair with current usage of ghtml and common-tags, as it creates the entire html tree with a single call, which is unrealistic. We should create another way to measure it by using components and rendering data lists.

arthurfiorette commented:
I'm not sure if someone is going to review these files, but if so, you can safely ignore everything inside packages/src or packages/html/*.js. No actual source code was changed since 8fb79e0a5e8b9b004291cc687c26b2e84d1f0e46 Only project setup and monorepo configurations were created.

JacopoPatroclo commented:
> What about creating another package just to have the benchmark code?
> 
> What do you think about the current benchmark? I'm not sure it they represent real usecase of this library. It would be awesome if we could copy a large website with components, [].map's, object styles and related stuff to compare in a real world scenario.

> Also, our current benchmark isn't fair with current usage of ghtml and common-tags, as it creates the entire html tree with a single call, which is unrealistic. We should create another way to measure it by using components and rendering data lists.

Fist sorry for my delayed response. I was away for work.
So, I think that creating a common code for benchmarking is not the right approach. Each jsx framework has its way of writing jsx (see the use of `className` vs `class` for example). We should keep each renderer separate but still check their output match to one done by @kitajs/html.

Regarding the "real word benchmark" thing. We could try to strive towards real-world benchmarks but the aim should not make @kitajs/html numbers better. The aim should be to have a way to measure how fast is this approach on rendering HTML over time.
For sure template libraries (common-tags, ghtml) have a performance advantage but they are missing all good developer experience around jsx. That should be the focus in my opinion for this library. Performance is important but dev-exp is essential.
The thing that I'm saying is, that we should put more effort into improving the developer experience around this library than making it the fastest HTML template engine out there.
Regardless of your opinion on this, I will still keep contributing and using this lib because I believe in the value that provides.

JacopoPatroclo commented:
@arthurfiorette I've pulled and rebased the current branch with the changes from the master to solve the conflict.

arthurfiorette commented:
> We could try to strive towards real-world benchmarks but the aim should not make @kitajs/html numbers better.

Honestly, I didn't tried to improve the kitajs/html numbers. Previous benchmarks were unfair to any JSX tool because it template benchs ware a single giant function with a single html\`\`` function call and no one creates web pages containing anything from the `html` tag to the innermost `p` tag in a single template.

For example, taking the below benchmark:

https://github.com/kitajs/html/blob/2cf8bd53eb7546435f70660b48d85dcde5b4757c/benchmark/renderers/many-components.tsx#L663

it consists of a string of 1477 characters, a variable `name` and a string of 3416 characters being concatenated multiple times, meanwhile kita was being called 314 times, a jsx call per div created.

I know this is what happens at the end, but no one creates webpages with such giant html\`\` templates, its awful to maintain and develop. This is why I created a RealWorldPage, containing components, list mapping and many attributes.

I added a benchmark's readme clarifying it and added a note in the performance readme.

![image](https://github.com/kitajs/html/assets/47537704/3ede3f3c-b7c2-4722-b439-930b6eb95839)

![image](https://github.com/kitajs/html/assets/47537704/2a940cd6-e795-473e-b0e8-45122bc38a97)

I'd love if you could review my grammar :)

arthurfiorette commented:
Also, thanks for your awesome work in this upcoming PR. Would you like to become a member of kitajs?

JacopoPatroclo commented:
> > We could try to strive towards real-world benchmarks but the aim should not make @kitajs/html numbers better.
> 
> Honestly, I didn't tried to improve the kitajs/html numbers. Previous benchmarks were unfair to any JSX tool because it template benchs ware a single giant function with a single html```function call and no one creates web pages containing anything from the`html`tag to the innermost`p` tag in a single template.
> 
> For example, taking the below benchmark:
> 
> https://github.com/kitajs/html/blob/2cf8bd53eb7546435f70660b48d85dcde5b4757c/benchmark/renderers/many-components.tsx#L663
> 
> it consists of a string of 1477 characters, a variable `name` and a string of 3416 characters being concatenated multiple times, meanwhile kita was being called 314 times, a jsx call per div created.
> 
> I know this is what happens at the end, but no one creates webpages with such giant html`` templates, its awful to maintain and develop. This is why I created a RealWorldPage, containing components, list mapping and many attributes.
> 
> I added a benchmark's readme clarifying it and added a note in the performance readme.
> 
> ![image](https://private-user-images.githubusercontent.com/47537704/309302205-3ede3f3c-b7c2-4722-b439-930b6eb95839.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkzNzI1OTksIm5iZiI6MTcwOTM3MjI5OSwicGF0aCI6Ii80NzUzNzcwNC8zMDkzMDIyMDUtM2VkZTNmM2MtYjdjMi00NzIyLWI0MzktOTMwYjZlYjk1ODM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzAyVDA5MzgxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4ZWE1YTFhOGJmMWNkN2UwZTczNmRiMGU3YWFjZDlhYjVjNDJlZTEzZWM3ZWRmMTRiNWM0YTljZDhmZjc1ODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5wYV8_9XCMLe2wlGMe4PpWH_sDDFV34ZO2ujDNz_jnU)
> 
> ![image](https://private-user-images.githubusercontent.com/47537704/309302259-2a940cd6-e795-473e-b0e8-45122bc38a97.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkzNzI1OTksIm5iZiI6MTcwOTM3MjI5OSwicGF0aCI6Ii80NzUzNzcwNC8zMDkzMDIyNTktMmE5NDBjZDYtZTc5NS00NzNlLWIwZTgtNDUxMjJiYzM4YTk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzAyVDA5MzgxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyNWIwY2M2MWNlZDdmODllMTQ0NzkxYTk2M2E1MjA2MWU3Yzc0MWI4MDZlYjk4ZDc3YjE5NDNmMWU1NzUyNTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bsT_xLlHww8qse5f0QHVR5IvlPJdaELZuvDfOhDnGpw)
> 
> I'd love if you could review my grammar :)

I wrote my comment before seeing the updated benchmarks. I believe your assumptions are correct. Now the benchmarks reflect more on what a real-world use of a templating library looks like. But, in my opinion, we are still falling short of showing how performant using suspense is in certain cases; moreover, the fact that some of those "competitors" don't support it. I will try to come up with something, in the future after we close this PR, to highlight better this aspect.

arthurfiorette commented:
> Yes, for sure!

Invite sent :)

arthurfiorette commented:
>  I will try to come up with something, in the future after we close this PR, to highlight better this aspect.

Maybe these benchmarks should only measure the overhead Suspense and renderToString brings to it. Something like rendering a 1s sleep suspense ends up in 1.2s, then .2s was the overhead.

arthurfiorette commented:
I've added changesets, before commiting any feature, please run `pnpm changeset`, choose all changed packages by your feature and add a release note to it. Changesets will automatically bump versions, create a GH release and a changelog.md

arthurfiorette commented:
Tasklist:

- [x] Update links and references inside all package jsons.
- [ ] Ensure tests are working on all packages
- [ ] Test CI/CI
- [x] Test changeset
- [x] Update docs and reference links
- [ ] Do a simple publish --dry-run to ensure all files are being bundled correctly
- [ ] See if there's no missing or left configuration files
- [x] Update monorepo readme's
- [ ] Create a CONTRIBUTING guide to teach how to use `pnpm changeset` before commit.

---

Regarding `ts-html-plugin` I'm not sure what's going on with these tests, but its probably related to the @kitajs/html types, as when I change `"@kitajs/html": "workspace:^"` (linked locally) to `"@kitajs/html": "3.1.1"` (downloaded from npm), tests starts to pass.

---

JacopoPatroclo commented:
> Tasklist:
> 
> * [x]  Update links and references inside all package jsons.
> * [x]  Ensure tests are working on all packages
> * [ ]  Test CI/CI
> * [x]  Test changeset
> * [x]  Update docs and reference links
> * [ ]  Do a simple publish --dry-run to ensure all files are being bundled correctly
> * [ ]  See if there's no missing or left configuration files
> * [x]  Update monorepo readme's
> * [ ]  Create a CONTRIBUTING guide to teach how to use `pnpm changeset` before commit.
> 
> Regarding `ts-html-plugin` I'm not sure what's going on with these tests, but its probably related to the @kitajs/html types, as when I change `"@kitajs/html": "workspace:^"` (linked locally) to `"@kitajs/html": "3.1.1"` (downloaded from npm), tests starts to pass.

I've figured out the issue related `ts-html-plugin` tests not passing using the kitajs/html workspace version. Debugging the ts-server was a pain. Hope the fix is good enough. The issue was a mix of things, starting from how pnpm links packages to how typescript resolves them in different ways depending on the status of the file system.

JacopoPatroclo commented on packages/ts-html-plugin/test/unsafe-tags.test.ts:
```diff
@@ -11,7 +11,6 @@ it('Detect xss prone usage', async () => {
       <>
         <div>
         <div>{html}</div>
-        <div>{object}</div>
```
@arthurfiorette I'm not sure about this, empty Objects are no longer supported by `kitajs/html`, right? Should we live this line by adding an expected ts error or should we, as I did, remove the line completely?

arthurfiorette replied:
Hmm, but the error thrown in this line is not from the ts-html-plugin and just a type error from typescript.

JacopoPatroclo replied:
Yep, that's correct. So we should not test this behavior. I'll leave it like this.

arthurfiorette commented:
> I've figured out the issue related `ts-html-plugin` tests not passing using the kitajs/html workspace version. Debugging the ts-server was a pain. Hope the fix is good enough.

For the love of god thank you!!! ts-server is a MAJOR pain.

arthurfiorette commented:
I'll review it asaic. In the meantime, [this article have been sent to me regarding <self /> closing tags](https://jakearchibald.com/2023/against-self-closing-tags-in-html/). Should we also remove it from kitajs/html?

arthurfiorette commented:
> In the meantime, [this article have been sent to me regarding closing tags](https://jakearchibald.com/2023/against-self-closing-tags-in-html/). Should we also remove it from kitajs/html?

React still uses self closing tags, so let's just keep it in this way.

arthurfiorette commented:
I think were ready to merge. @JacopoPatroclo is there anything else missing?

JacopoPatroclo commented:
> I think were ready to merge. @JacopoPatroclo is there anything else missing?

@arthurfiorette  I think we should do a release candidate of the packages to test them inside some real projects. Just to make sure that they all are working properly. Can you manage to do that? I will test them on some of my projects as well. After that, we can close this PR.

If you like I could add Nx to enable caching and task orchestration to the monorepo. Of course, I will do it in a separate PR. Is there a place where we can discuss stuff like this outside of PR's comments?

arthurfiorette commented:
> I think we should do a release candidate of the packages to test them inside some real projects.

Me too, I'll just merge this to main because the changeset is set up there...

> If you like I could add Nx to enable caching and task orchestration to the monorepo.

Pnpm already does caching, task orchestration isn't needed for now as no complicated build step is required. Changesets already does releasing stuff too, so I think its enough for now.

arthurfiorette commented:
Anyways, thanks for the big effort you've put in here :)

arthurfiorette commented:
> Is there a place where we can discuss stuff like this outside of PR's comments?

Yes, add me on discord: @arthurfiorette
